File: prototype_min/build_dataset.py
#!/usr/bin/env python3
import json
from datetime import datetime, timezone
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
SAMPLE_DIR = ROOT / "0_platform_choice" / "data" / "samples" / "2026-04-07-snap-sponsored-content"
OUT_DIR = Path(__file__).resolve().parent / "data"
OUT_FILE = OUT_DIR / "normalized.json"


def read_json(path: Path):
    with path.open("r", encoding="utf-8") as f:
        return json.load(f)


def classify_text(text: str):
    t = (text or "").lower()
    if any(k in t for k in ["podcast", "playlist", "music", "spotify", "hörbuch", "audio"]):
        return ("Audio", "Streaming")
    if any(k in t for k in ["crypto", "trading", "investment", "invest"]):
        return ("Finance", "Investment")
    if any(k in t for k in ["tv", "show", "movie", "serie"]):
        return ("Entertainment", "Video")
    return ("Unknown", "Unknown")


def detect_risk(text: str, url: str):
    blob = ((text or "") + " " + (url or "")).lower()
    flags = []
    for term in ["crypto", "trading", "investment", "free money", "giveaway", "urgent", "scam"]:
        if term in blob:
            flags.append(term)
    return flags


def safe_int(v, default=0):
    try:
        return int(v)
    except Exception:
        return default


def load_ads_from_files():
    candidates = [
        "ads_search_spotify_de_retry.json",
        "ads_search_nike_de_local.json",
        "ads_search_nike_de.json",
    ]
    merged = {"ad_previews": []}
    used = []
    for name in candidates:
        path = SAMPLE_DIR / name
        if not path.exists():
            continue
        payload = read_json(path)
        if payload.get("request_status") == "SUCCESS":
            merged["ad_previews"].extend(payload.get("ad_previews", []))
            used.append(name)
    return merged, used


def main():
    ads, ads_source_files = load_ads_from_files()
    sponsored = read_json(SAMPLE_DIR / "sponsored_content_pages_1_2.json")

    now = datetime.now(timezone.utc)
    ads_rows = []
    sponsored_rows = []

    for item in ads.get("ad_previews", []):
        ad = item.get("ad_preview")
        if not ad:
            continue
        start_date = ad.get("start_date")
        dt = None
        if start_date:
            try:
                dt = datetime.fromisoformat(start_date.replace("Z", "+00:00"))
            except Exception:
                dt = None
        running_days = (now - dt).days if dt else None
        impressions_total = safe_int(ad.get("impressions_total"))
        cpm_low = 4.0
        cpm_high = 9.0
        est_spend_low = round(impressions_total / 1000 * cpm_low, 2)
        est_spend_high = round(impressions_total / 1000 * cpm_high, 2)
        headline = ad.get("headline", "")
        url = (ad.get("web_view_properties") or {}).get("url", "")
        category, sub_category = classify_text(headline + " " + url)
        risk_flags = detect_risk(headline, url)

        ads_rows.append(
            {
                "id": ad.get("id"),
                "platform": "Snapchat",
                "advertiser": ad.get("paying_advertiser_name") or "Unknown",
                "brand": ad.get("brand_name") or ad.get("profile_name") or "Unknown",
                "headline": headline,
                "cta": ad.get("call_to_action"),
                "start_date": start_date,
                "running_days": running_days,
                "impressions_total": impressions_total,
                "impressions_primary_country": safe_int((ad.get("impressions_map") or {}).get("de")),
                "est_spend_low_eur": est_spend_low,
                "est_spend_high_eur": est_spend_high,
                "performance_per_spend_proxy": round(
                    impressions_total / max((est_spend_low + est_spend_high) / 2, 1), 2
                ),
                "category": category,
                "sub_category": sub_category,
                "unknown_category": category == "Unknown",
                "risk_flags": risk_flags,
                "risk_score": min(len(risk_flags) * 2, 10),
                "media_type": ad.get("top_snap_media_type") or "Unknown",
                "creative_url": ad.get("top_snap_media_download_link"),
                "landing_url": url,
                "review_status": ad.get("review_status"),
            }
        )

    sponsored_payload = sponsored
    if isinstance(sponsored, dict):
        sponsored_payload = [sponsored]

    for page in sponsored_payload:
        for item in page.get("ad_previews", []):
            sc = item.get("sponsored_content_preview")
            if not sc:
                continue
            creator = sc.get("creator_name", "Unknown")
            sponsor = sc.get("sponsor_name", "") or "Unknown"
            category, sub_category = classify_text(f"{creator} {sponsor}")
            sponsored_rows.append(
                {
                    "creator_name": creator,
                    "sponsor_name": sponsor,
                    "content_type": sc.get("content_type"),
                    "content_url": sc.get("content_url"),
                    "thumbnail_url": sc.get("thumbnail_url"),
                    "category": category,
                    "sub_category": sub_category,
                }
            )

    sponsor_counts = {}
    creator_counts = {}
    for row in sponsored_rows:
        sponsor = row.get("sponsor_name") or "Unknown"
        creator = row.get("creator_name") or "Unknown"
        sponsor_counts[sponsor] = sponsor_counts.get(sponsor, 0) + 1
        creator_counts[creator] = creator_counts.get(creator, 0) + 1

    top_sponsors = [
        {"sponsor_name": name, "count": count}
        for name, count in sorted(sponsor_counts.items(), key=lambda x: x[1], reverse=True)[:10]
    ]
    top_creators = [
        {"creator_name": name, "count": count}
        for name, count in sorted(creator_counts.items(), key=lambda x: x[1], reverse=True)[:10]
    ]

    advertisers = {}
    for r in ads_rows:
        key = r["advertiser"]
        advertisers.setdefault(key, {"impressions": 0, "est_low": 0.0, "est_high": 0.0, "ads": 0})
        advertisers[key]["impressions"] += r["impressions_total"]
        advertisers[key]["est_low"] += r["est_spend_low_eur"]
        advertisers[key]["est_high"] += r["est_spend_high_eur"]
        advertisers[key]["ads"] += 1

    leaderboard = [
        {
            "advertiser": k,
            "ads": v["ads"],
            "impressions": v["impressions"],
            "est_spend_low_eur": round(v["est_low"], 2),
            "est_spend_high_eur": round(v["est_high"], 2),
        }
        for k, v in advertisers.items()
    ]
    leaderboard.sort(key=lambda x: x["impressions"], reverse=True)

    output = {
        "generated_at_utc": now.isoformat(),
        "notes": {
            "spend_model": "Very rough proxy from impressions using fixed CPM range 4-9 EUR.",
            "performance_model": "Impressions per estimated euro midpoint.",
            "warning": "Prototype only. Do not treat spend as exact.",
            "ads_source_files_used": ads_source_files,
        },
        "kpis": {
            "ads_count": len(ads_rows),
            "sponsored_content_count": len(sponsored_rows),
            "total_impressions": sum(x["impressions_total"] for x in ads_rows),
            "est_spend_low_total_eur": round(sum(x["est_spend_low_eur"] for x in ads_rows), 2),
            "est_spend_high_total_eur": round(sum(x["est_spend_high_eur"] for x in ads_rows), 2),
            "unknown_category_ratio": round(
                (sum(1 for x in ads_rows if x["unknown_category"]) / max(len(ads_rows), 1)) * 100, 1
            ),
        },
        "sponsored_kpis": {
            "sponsored_rows": len(sponsored_rows),
            "unique_creators": len(creator_counts),
            "unique_sponsors": len(sponsor_counts),
            "top_sponsors": top_sponsors,
            "top_creators": top_creators,
            "scope_note": "Represents sponsored/commercial content endpoint coverage, not full organic platform feed.",
        },
        "leaderboard": leaderboard,
        "ads": ads_rows,
        "sponsored_content": sponsored_rows,
    }

    OUT_DIR.mkdir(parents=True, exist_ok=True)
    with OUT_FILE.open("w", encoding="utf-8") as f:
        json.dump(output, f, ensure_ascii=False, indent=2)

    print(f"Wrote {OUT_FILE}")


if __name__ == "__main__":
    main()

File: app/build_dataset.mjs
#!/usr/bin/env node
/**
 * Builds app/data/normalized.json from DuckDB (db/rav.db).
 *
 * Prereqs: DuckDB CLI (`brew install duckdb`) and a populated database
 *          (`duckdb db/rav.db < db/init.sql` from repo root).
 */
import { execSync } from "node:child_process";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dirname, "..");
const DB_PATH = path.join(ROOT, "db", "rav.db");
const OUT_DIR = path.join(__dirname, "data");
const OUT_FILE = path.join(OUT_DIR, "normalized.json");

// ── DuckDB helper ──────────────────────────────────────────

function query(sql) {
  const out = execSync(`duckdb "${DB_PATH}" -json`, {
    input: sql + "\n",
    encoding: "utf8",
    maxBuffer: 100 * 1024 * 1024,
  });
  return JSON.parse(out || "[]");
}

// ── Enrichment (JS-side) ──────────────────────────────────

const CATEGORY_RULES = [
  [["podcast", "playlist", "music", "spotify", "hörbuch", "audio"], "Audio", "Streaming"],
  [["crypto", "trading", "investment", "invest"], "Finance", "Investment"],
  [["tv", "show", "movie", "serie"], "Entertainment", "Video"],
  [["nike", "adidas", "puma", "jordan", "new balance", "asics", "vans", "salomon", "crocs", "birkenstock", "dr. martens", "converse"], "Fashion", "Footwear"],
  [["shein", "zara", "h&m", "boohoo", "asos", "mango", "cos", "arket", "only", "jack & jones", "kiabi", "prettylittlething", "fila", "champion", "superdry", "g-star", "lacoste", "tommy hilfiger", "calvin klein", "hugo boss", "ralph lauren", "diesel", "guess", "patagonia", "lululemon", "triumph", "under armour", "foot locker", "zalando", "house"], "Fashion", "Apparel"],
  [["sephora", "dior", "chanel", "gucci", "louis vuitton", "givenchy", "prada", "versace", "armani", "fendi", "bottega veneta", "valentino", "celine", "kenzo", "cartier", "tiffany", "rolex", "omega", "pandora", "hermes"], "Fashion", "Luxury"],
  [["mac", "nyx", "clinique", "l'oreal", "charlotte tilbury", "nars", "clarins", "kylie cosmetics", "olay", "dove", "axe", "braun", "ghd", "coty", "essence", "paco rabanne"], "Beauty", "Cosmetics"],
  [["temu", "coach", "oakley", "specsavers", "galeries lafayette", "mammut", "canada goose"], "Retail", "General"],
  [["game", "gaming", "esport"], "Entertainment", "Gaming"],
];

const RISK_TERMS = ["crypto", "trading", "investment", "free money", "giveaway", "urgent", "scam"];

function classifyText(text) {
  const t = (text || "").toLowerCase();
  for (const [keywords, cat, sub] of CATEGORY_RULES) {
    if (keywords.some((k) => t.includes(k))) return [cat, sub];
  }
  return ["Unknown", "Unknown"];
}

function detectRisk(text, url) {
  const blob = `${text || ""} ${url || ""}`.toLowerCase();
  return RISK_TERMS.filter((term) => blob.includes(term));
}

// ── Main ──────────────────────────────────────────────────

function main() {
  if (!fs.existsSync(DB_PATH)) {
    throw new Error(`DuckDB not found at ${DB_PATH}. Run: duckdb db/rav.db < db/init.sql`);
  }

  // 1. Ads from brand_ads_fashion (all 4.5k+ rows, 23 EU countries)
  const rawAds = query(`
    SELECT
      ad_id                         AS id,
      COALESCE(paying_advertiser_name, 'Unknown') AS advertiser,
      COALESCE(NULLIF(brand_name_api,''), NULLIF(profile_name,''), brand, 'Unknown') AS brand,
      COALESCE(headline, '')        AS headline,
      call_to_action                AS cta,
      start_date::VARCHAR           AS start_date,
      CASE WHEN start_date IS NOT NULL
           THEN DATE_DIFF('day', start_date::DATE, CURRENT_DATE)
      END                           AS running_days,
      COALESCE(CAST(impressions_total AS INTEGER), 0) AS impressions_total,
      COALESCE(top_snap_media_type, 'Unknown') AS media_type,
      top_snap_media_download_link  AS creative_url,
      COALESCE(web_view_properties.url, '') AS landing_url,
      review_status,
      country
    FROM brand_ads_fashion
  `);

  const CPM_LOW = 4.0;
  const CPM_HIGH = 9.0;

  const adsRows = rawAds.map((ad) => {
    const impr = ad.impressions_total || 0;
    const estLow = Math.round((impr / 1000) * CPM_LOW * 100) / 100;
    const estHigh = Math.round((impr / 1000) * CPM_HIGH * 100) / 100;
    const blob = `${ad.headline} ${ad.landing_url} ${ad.brand} ${ad.advertiser}`;
    const [category, sub_category] = classifyText(blob);
    const riskFlags = detectRisk(ad.headline, ad.landing_url);
    return {
      id: ad.id,
      platform: "Snapchat",
      advertiser: ad.advertiser,
      brand: ad.brand,
      headline: ad.headline,
      cta: ad.cta,
      start_date: ad.start_date,
      running_days: ad.running_days,
      impressions_total: impr,
      est_spend_low_eur: estLow,
      est_spend_high_eur: estHigh,
      performance_per_spend_proxy:
        Math.round((impr / Math.max((estLow + estHigh) / 2, 1)) * 100) / 100,
      category,
      sub_category,
      unknown_category: category === "Unknown",
      risk_flags: riskFlags,
      risk_score: Math.min(riskFlags.length * 2, 10),
      media_type: ad.media_type,
      creative_url: ad.creative_url,
      landing_url: ad.landing_url,
      review_status: ad.review_status,
      country: ad.country,
    };
  });

  // 2. Sponsored content KPIs (computed in DuckDB over full 230k+ row table)
  const [sponsoredCounts] = query(`
    SELECT
      COUNT(*)                                   AS sponsored_rows,
      COUNT(DISTINCT creator_name)               AS unique_creators,
      COUNT(DISTINCT NULLIF(sponsor_name, ''))   AS unique_sponsors
    FROM sponsored_content
  `);

  const topSponsors = query(`
    SELECT sponsor_name, COUNT(*) AS count
    FROM sponsored_content
    WHERE sponsor_name != ''
    GROUP BY sponsor_name
    ORDER BY count DESC
    LIMIT 10
  `);

  const topCreators = query(`
    SELECT creator_name, COUNT(*) AS count
    FROM sponsored_content
    GROUP BY creator_name
    ORDER BY count DESC
    LIMIT 10
  `);

  const sponsoredSample = query(`
    SELECT creator_name, sponsor_name, content_type, content_url, thumbnail_url
    FROM sponsored_content
    WHERE sponsor_name != ''
    USING SAMPLE 200
  `);

  // 3. Aggregate ad-level KPIs
  const advertisers = {};
  for (const r of adsRows) {
    const key = r.advertiser;
    if (!advertisers[key]) advertisers[key] = { impressions: 0, est_low: 0, est_high: 0, ads: 0 };
    advertisers[key].impressions += r.impressions_total;
    advertisers[key].est_low += r.est_spend_low_eur;
    advertisers[key].est_high += r.est_spend_high_eur;
    advertisers[key].ads += 1;
  }

  const leaderboard = Object.entries(advertisers)
    .map(([advertiser, v]) => ({
      advertiser,
      ads: v.ads,
      impressions: v.impressions,
      est_spend_low_eur: Math.round(v.est_low * 100) / 100,
      est_spend_high_eur: Math.round(v.est_high * 100) / 100,
    }))
    .sort((a, b) => b.impressions - a.impressions);

  const now = new Date();
  const output = {
    generated_at_utc: now.toISOString(),
    notes: {
      spend_model: "Very rough proxy from impressions using fixed CPM range 4-9 EUR.",
      performance_model: "Impressions per estimated euro midpoint.",
      warning: "Prototype only. Do not treat spend as exact.",
      data_source: "DuckDB (db/rav.db) — brand_ads_fashion + sponsored_content tables",
    },
    kpis: {
      ads_count: adsRows.length,
      sponsored_content_count: sponsoredCounts.sponsored_rows,
      total_impressions: adsRows.reduce((s, x) => s + x.impressions_total, 0),
      est_spend_low_total_eur:
        Math.round(adsRows.reduce((s, x) => s + x.est_spend_low_eur, 0) * 100) / 100,
      est_spend_high_total_eur:
        Math.round(adsRows.reduce((s, x) => s + x.est_spend_high_eur, 0) * 100) / 100,
      unique_brands: new Set(adsRows.map((r) => r.brand)).size,
      unique_countries: new Set(adsRows.map((r) => r.country)).size,
      unknown_category_ratio:
        Math.round(
          ((adsRows.filter((x) => x.unknown_category).length / Math.max(adsRows.length, 1)) * 100) * 10
        ) / 10,
    },
    sponsored_kpis: {
      sponsored_rows: sponsoredCounts.sponsored_rows,
      unique_creators: sponsoredCounts.unique_creators,
      unique_sponsors: sponsoredCounts.unique_sponsors,
      top_sponsors: topSponsors,
      top_creators: topCreators,
      scope_note:
        "Represents sponsored/commercial content endpoint coverage, not full organic platform feed.",
    },
    leaderboard,
    ads: adsRows,
    sponsored_content: sponsoredSample,
  };

  fs.mkdirSync(OUT_DIR, { recursive: true });
  fs.writeFileSync(OUT_FILE, JSON.stringify(output, null, 2), "utf8");

  const stats = [
    `${adsRows.length} ads (${new Set(adsRows.map((r) => r.brand)).size} brands, ${new Set(adsRows.map((r) => r.country)).size} countries)`,
    `${sponsoredCounts.sponsored_rows} sponsored rows (${sponsoredCounts.unique_creators} creators)`,
  ].join(", ");
  console.log(`Wrote ${OUT_FILE} — ${stats}`);
}

main();
